elastic/kibana
# Update alert documents when the write index changes

Fixes https://github.com/elastic/kibana/issues/110519

### 📝 Summary
The proposed solution in the above ticket to remove `require_alias: true` didn't work, because the mappings wouldn't be installed correctly. This is because of the current logic where resources are installed only when the bulk operation fails https://github.com/elastic/kibana/blob/master/x-pack/plugins/rule_registry/server/rule_data_client/rule_data_client.ts#L139. If the flag is missing and we would like to index some new data, resources wouldn't be installed and alerts table wouldn't render any data. 

So the fix was to:
- keep the `require_alias`: true
- and disable it only when updating a document

```
await ruleDataClient.getWriter().bulk({
      body: allEventsToIndex.flatMap(({ event, indexName }) => [
        indexName
          ? { index: { _id: event[ALERT_UUID]!, _index: indexName, require_alias: false } }
          : { index: { _id: event[ALERT_UUID]! } },
        event,
      ]),
    });
```


### How to test
For both scenarios here's the command you can use in Dev tools to get the alerts that are being indexed
```
GET .alerts-observability*/_search
{
  "fields": [
    "*"
  ],
  "_source": false,
  "sort": [
    {
      "kibana.alert.start": {
        "order": "desc" 
      }
    }
  ]
}
```

**Scenario 1** (an ongoing alert gets written in old index after rollover)
- Create a new rule and generate some data that should trigger an alert
- Verify one new alert is written in the correct index `.internal.alerts-observability.logs.alerts-default-000001`
- Wait for the next trigger of the alert and verify that alert is updated and no new alert is created
- In Devtools do a rollover `POST .alerts-observability.logs.alerts-default/_rollover`
- Verify you can see two indices `GET .alerts-observability.logs.alerts-default`
- Wait for the alert to trigger again
- Verify the new alert is still written in old new index `.internal.alerts-observability.logs.alerts-default-000001`

**Scenario 2** (an ongoing alert **SHOULD** be written in the old index after rollover and after a new rule type was created)
- Create a new rule and generate some data that should trigger an alert
- Verify one new alert is written in the correct index `.internal.alerts-observability.logs.alerts-default-000001`
- Wait for the next trigger of the alert and verify that alert is updated and no new alert is created
- In Devtools do a rollover `POST .alerts-observability.logs.alerts-default/_rollover`
- Verify you can see two indices `GET .alerts-observability.logs.alerts-default`
- Create a new rule and wait for the new alert to trigger
- Verify the new alert is written in the new index `.internal.alerts-observability.logs.alerts-default-000002`
- Verify the old alert is written in the old index `.internal.alerts-observability.logs.alerts-default-000001` => ~~We just spotted a bug 🐞 🐛 , where ILM policy deleted the old indices after rollover and the old alert got indexed in `.internal.alerts-observability.logs.alerts-default-000002` instead~~ https://github.com/elastic/kibana/issues/111029
